Syslog Support

sbchisholm opened the pull request:
- Adds support for multiple logging output.
- Configure log level via string (i.e.: `"INFO"` or `"DEBUG"` etc.).
- NOTE: Breaks the log level numbers (the log level enum is now in line with syslog).

kmsquire commented on src/Logging.jl:
```diff
@@ -1,19 +1,21 @@
 end
 
-@enum LogLevel DEBUG INFO WARNING ERROR CRITICAL OFF
+@enum LogLevel OFF=-1 CRITICAL=2 ERROR WARNING INFO=6 DEBUG
```
This is better, and even though breaking, probably okay, as long as no one was depending on the order outside of this module.  Will probably do a minor version bump.

kmsquire commented on src/Logging.jl:
```diff
@@ -21,23 +23,66 @@ catch
     name::String
     level::LogLevel
-    output::IO
+    output::Array{LogOutput,1}
```
Can you change this to:
```julia
type Logger{L<:LogOutput}
...
    output::Array{L,1}  # or output::Vector{L}
...
```

Otherwise, extra code is generated to determine the actual element type every time you log something.

(It should probably have been like this before, but I wrote this code well before I understood the the consequences of abstract types for member variables.)

kmsquire replied:
Actually, never mind--this would prevent you from having different types of output for the same logger.  (Which might have been obvious to you.)

kmsquire commented on src/Logging.jl:
```diff
@@ -21,23 +23,66 @@ catch
+    # syslog needs a timestamp in the form: YYYY-MM-DDTHH:MM:SS-TZ:TZ
+    timestamp = string(strftime("%Y-%m-%dT%H:%M:%S",t), strftime("%z",t)[1:end-2], ":", strftime("%z",t)[end-1:end])
+    log_msg = string("<", (uint16(syslog.facility) << 3) + uint16(level), ">1 ", timestamp, " ", syslog.machine, " ", syslog.user, " - - - ", level, ":", logger_name,":", msg...)
+   send(syslog.socket, syslog.ip, syslog.port, log_msg)
```
3 spaces.  Since this is the only issue I see, I'll merge later today even if you don't fix it.

kmsquire commented:
Looks really good--thanks!

sbchisholm commented:
Thanks for the feedback.  It's also worth noting this only adds support for logging to syslog via UDP.  Syslog also supports logging to a local socket and over TCP (something to look into adding in the future).

sbchisholm commented:
@kmsquire I recently updated this PR, fixed a few julia 0.4 deprecation warnings and added a max message length attribute to the SysLog type, this allows you to truncate log messages.  I was finding that my application was crashing on some large debug messages when trying to send them over the `UDPSocket`.

DanielArndt commented:
@kmsquire Any plans to merge this in? 

We're going to be keeping a fork at https://github.com/analyzere/Logging.jl so I have a vested interest in preventing these two from diverging too much :)

kmsquire commented:
Thanks for the bump.  I've been working my way through issues in other packages, and haven't spent a lot of love here.

DanielArndt commented:
No problem, I've seen the other things you've been doing, you're very busy. Thanks!

kmsquire commented:
@sbchisholm @DanielArndt I've added you both as collaborators here, so if there other things that you see that need attention, I'd love the help.

My hopes/plans:
1. move this repo to JuliaLang
2. integrate the basic framework into Julia itself, and have this package and/or others just provide extra features.  (https://github.com/JuliaLang/julia/issues/13515)
3. borrow ideas/code from LumberJack.jl (which is probably pretty good, but I can't get myself past the naming scheme)

It seems that people use this package, so I think the first item is reasonable so that at least more people have access to fix things.  That said, just moving it to JuliaLang doesn't guarantee that.

For the others, I don't know if/when I will find the time. Like a lot of people, I'm a volunteer here, and my availability ebbs and flows.  I'm also about to become a first-time father, so I should probably take care of at least the first item pretty soon... ;-)
